README
Net::DNS::SEC - DNSSEC extensions to Net::DNS
=============================================


IMPORTANT UPGRADE NOTE.

Please always read the Changes file. It is accessible
through: http://search.cpan.org/src/OLAF/Net-DNS-SEC/Changes

	
1. GENERAL INFO.

  This module implements DNS Resource Record types that are relevant
  for DNSSEC operations. This package implements RFC4034
  "Resource Records for the DNS Security Extensions" and RFC2931 "DNS
  Request and Transaction Signatures ( SIG(0)s )"

  RFC 3225 (Indicating Resolver Support of DNSSEC) support has been
  integrated into Net::DNS.

  The extensions provide the following additional features on top of
  the Net::DNS package (http://www.net-dns.org/). 

  - DS, DNSKEY, RRSIG, NSEC and NSEC3 records. These are all implemented
    as Net::DNS RR objects.  The cryptography has been implemented using
    Crypt::OpenSSL::DSA, Crypt::OpenSSL::RSA and related modules.

  The reason for not making these RRs a part of the regular Net::DNS
  distribution is that they rely on crypto modules that do not easily
  port.

  Also see the 'pod' documentation in: 
       Net::DNS::RR::DNSKEY
       Net::DNS::RR::DS
       Net::DNS::RR::NSEC
       Net::DNS::RR::NSEC3
       Net::DNS::RR::RRSIG
       Net::DNS::SEC::Private
  and for the use of SIG0 see:
       Net::DNS::Packet (sign_sig0)
       Net::DNS::RR::KEY
       Net::DNS::RR::SIG
       Net::DNS::SEC::Private


  On top of the RR classes defined above we also provide a class for
  handling keysets. Keysets are administrative files used by the BIND
  tools for keymaintenance tasks.  Net::DNS::Keyset provides an abstract
  interface for doing fun things with them.

2. Dependencies

  This package relies on Net-DNS version 0.69 or higher (also see CHANGES)
  which has features specifically designed to support Net::DNS::SEC.

  Among other straightforward routines you will need to have
  Crypt::OpenSSL::DSA and Crypt::OpenSSL::RSA installed.

  The package was initially developed on FreeBSD 4.5 with perl5.6.1, 
  as of version 0.11 development and testing where done on 
  MacOS 10.2.6 with perl, v5.8.0 built for darwin.


3. Install

  Use CPAN.

  perl -MCPAN -e shell;
  cpan> install Net::DNS::SEC
  

  Alternatively install this package manually:
    tar -xvzf Net-DNS-SEC-??.tar.gz
    cd  Net-DNS-SEC-??.tar.gz
    perl Makefile.PL
    make 
    make test
    make install

  'perl Makefile.PL' will test if all dependencies are in place. 
  'make test' will do some tests that should all succeed once all
  dependencies are satisfied. Please report failures during the test 
  phase.


4. Maintenance

  Please use the CPAN request tracker if you find bugs in the
  extensions or have DNSSEC feature request. The request tracker is 
  available through:

   https://rt.cpan.org/NoAuth/ReportBug.html?Queue=Net-DNS-SEC
  
  Net::DNS contains a number of hooks for Net::DNS::SEC in the
  Net::DNS package. If you have problems with Net::DNS in general see:
  http://www.net-dns.org/

  A snapshot of the development code is available through subversion:
	http://www.net-dns.org/svn/net-dns-sec/trunk

  Note that the code on the trunk may not be functional.


5. Staying up to date.

  Announcements about Net::DNS::SEC will be done on 
  http://www.net-dns.org/blog/. An RSS feed is available.


6. Demos
  demo/getkeyset.pl - is a small demonstration program that will get
                      fetch the keyset for a specified domain, stores
                      them and prints out the DS RRs to STD out.
  demo/make-signed-keyset - A script contributed by Wes Griffin creates 
                      a self-signed keyset from a BIND public key file 
                      specified on the command line.

  demo/key2ds      -  reads the key data from STDIN and print the 
                      corresponding DS record on STDOUT. Contributed by 
                      Miek Gieben


7. History

  The modules in this package have somewhat of a history as a set of 
  patches to the original Net::DNS package. 

  Version 0.02 is the first distribution of the DNSSEC security
  for CPAN.  See the Changes file for more information.


8. Acknowledgements

  Thanks to the numerous folk that took an interest during
  development.  People that contributed to bug fixes should be
  acknowledged in the Changes file. If I forgot to mention you there
  it is not out of bad intention.

  Thanks to Chris Reinardt for putting the hooks in Net::DNS and suggesting
  some improvements to the package.
  Thanks Ian Robertson and T.J. Mather for their support with the
  Crypt::OpenSSL::[RSA|DSA] modules.

  



9. Author information

  Olaf M. Kolkman, RIPE NCC.
  support: olaf@net-dns.org


10. COPYRIGHT

Copyright Notice and Disclaimer

Copyright (c) 2001-2005 RIPE NCC.  Author Olaf M. Kolkman
<net-dns-sec@ripe.net>

All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the name of the author not be
used in advertising or publicity pertaining to distribution of the
software without specific, written prior permission.


THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE, INCLUDING
ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO EVENT SHALL
AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY
DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


Based on, and contains, code by Copyright (c) 1997-2001 Michael Fuhr.




------------------------------------------------------------------------------
$Id: README 1289 2015-01-05 10:08:59Z willem $


Changes
Revision history for Perl extension Net::DNS::SEC.


**** 0.22 [unreleased]

   Fix: rt.cpan.org #101184
   make siginception and sigexpiration available as time() values

   Fix: rt.cpan.org #101183
   wrong URL for blog in README

   Fix: rt.cpan.org #83031
   [RRSIG] lack of ECDSA support


***0.21 October 24, 2014

   Fix: rt.cpan.org #99250
   [RRSIG] validation fails when Signer's Name is upper case

   Fix: rt.cpan.org #99106
   Premature end of base64 data  (in 14-misc.t test script)


***0.20 August 15, 2014

   Fix: rt.cpan.org #97457
   !hex! error when parsing NSEC3 with empty salt


***0.19 Jun 6, 2014

   Remove inappropriate deprecation warning in DNSKEY.pm


***0.18 May 8, 2014

   Recode RR implementations to provide Net::DNS 0.69+ interface.

   Fix: rt.cpan.org #95034
   Failure to parse NSEC3PARAM record with null salt

   Fix: rt.cpan.org #81289
   Failure to handle GOST DS records


***0.17 November 29, 2013

   Fix: rt.cpan.org #90270
   NSEC3->covered() should use case-insensitive comparisons

   Fix: rt.cpan.org #79606
   Lower case zone-name part with DNSKEY::privatename

   Allow to specify algorithms with ::Private->new_rsa_private and
   ::Private->generate_rsa instead of assuming RSASHA1.

   Fix: rt.cpan.org #55621
   Specify license type (mit) in META.yml

   Fix: rt.cpan.org #60269
   Remove Digest::SHA1 prerequirement

   Fix: rt.cpan.org #62387 & #63273
   Typo fixes

   Fix: rt.cpan.org #62386
   Make Net::DNS::RR::DS::digtype method work

   Fix: rt.cpan.org #62385
   Do not compress Next Domain Name in NSEC rdata.

   Fix: rt.cpan.org #61104
   Spelling correction in DS.pm and fix of key2ds demo program

   Fix: rt.cpan.org #60185
   Make sure %main::SIG hash keeps its values when compiling Net::DNS::RR::SIG
   in perl versions before 5.14.  See also: rt.perl.org #76138

   Fix: rt.cpan.org #64552 and rt.cpan.org #79606
   Support for private-key-format v1.3

   Fix: rt.cpan.org #75892
   Do not canonicalize the "Next Domain Name" rdata field of a NSEC RR
   for draft-ietf-dnsext-dnssec-bis-updates-17 compliance.

   BUG FIX/FEATURE: validation of wildcard RRs now available. Duane
   Wessels is acknowledged for submitting the initial code on which
   this fix is based.

   FIX: case sensitivity of ownername during DS generation
        (Acknowledgements Jens Wagner)


***0.16 March 12, 2010

   Feature: KEY inherits DNSKEY
   This helps maintenance in one part of the code.

   Feature: keylength methode rt.cpan.org #53468
   Added keylength method for RSA and DSA
   Acknowledgements Hugo Salgado

   Fix: rt.cpan.org #51778
   Empty bitmap would cause error about undefined ARRAY in NSEC/NSEC3.
   Now the code will allow empty bitmaps gracefully    

   Feature: New Algorithm Support (rt.cpan.org #51092) 
   SHA2 algorithm support, including NSEC3 algorithm parameters updated
   Acknowledgement Jakob Shlyter

   Fix: rt.cpan.org #42089
   NSEC3 Algorithm support in NSEC3 broken
   patch by Wes Hardaker


***0.15 December 31, 2008


   Fix: digestbin not set when an empty value passed to hash.

   Feature: Added DLV (rfcc 4431). The RR object is simply a clone of
   the DS RR and inherits ... everything

   Feature: Added NSEC3 and NSEC3PARAM support (RFC5155). 
   This adds Mime::Base32 to the module dependency list. 
   The RR type was still experimental at that time and is maintained
   in Net::DNS::RR.

   Fix: Test script recognizes change in Time::Local. Note that
   Time::Local does not deal with dates beyond 03:14:07 UTC on
   Tuesday, 19 January 2038. Therefore this code has a year 2038 
   problem.

   Fix: DS create_from_hash now produces objects that can create
   wireformat.

   Other: minor changes to the debug statements
          added t/05-rr.t (and identified a couple of bugs using it)

   Fix: a few inconsistencies with respect to parsing of trailing dots.


   During development the test signatures generated with the BIND tools
   were re-generated in order to troubleshoot a bug that (most
   probably) was caused by a version incompatibility between Net::DNS
   and Net::DNS::SEC.  Before release the original test from the 0.14
   release were ran against this version too.
   

 

0.14 February 14, 2005

   FIX: The introducion of the keytag warning triggered a bug with RSAMD5
        keys, causing RSAMD5 keys not to be loaded. 



0.13 December 9, 2005 

   FEAT: rt.cpan.org 14588

        Added support for passing (a reference to) an array of keys to the
	RRSIG verify function.

   FIX/FEAT:
	The Net::DNS::SEC::Private function will for RSA based keys verify if
	the keytag in the filename is actually correct.
	Since at parsing the value of the DNSKEY RR flags is not known we
	test against the currently defined flag values 256 and 257.

	If we cannot find a keytag match a warning is printed and Private
	key generation fails

        This inconsistency was spotted by Jakob Shlyter.
	

   FEAT: Added support for SHA256 to the DS RR. Assigned the expected 
        digest type2 for SHA256 type hashes.
	
        Note that this makes the Net::DNS::SEC depend on Digest::SHA instead
        of Digest::SHA1.

        The default digest type is still set to 1.

        NB. The code makes assumptions about the IANA assignment of the 
            digest type. The assignment may change. Do not use SHA256 in 
	    production zones!!

   FIX: rt.cpan.org #15662

	Roy Arends noticed and patched the label counting did not ignore
        an initial asterisk label.

   FIX: Wes Hardaker noticed the default TTL values for created signatures to
        be different from the TTLs from the data that is being signed.

   FIX: Wes Hardaker reported there was a problem with validating
        RRsets that had ownernames with capitals.
	The fix depends on a fix in Net::DNS::RR that is available in
	version 0.53_03 or later of the Net::DNS distribution.


  FEAT: Propper dealing with mnemonics for algorithm and digest type
	added to DS


RR/DS.pm
package Net::DNS::RR::DS;

#
# $Id: DS.pm 1276 2014-10-19 06:02:40Z willem $
#
use vars qw($VERSION);
$VERSION = (qw$LastChangedRevision: 1276 $)[1];


use strict;
use base qw(Net::DNS::RR);

=head1 NAME

Net::DNS::RR::DS - DNS DS resource record

=cut


use integer;

use warnings;
use Carp;

use constant BABBLE => eval { require Digest::BubbleBabble; };

eval { require Digest::SHA };		## optional for simple Net::DNS RR
eval { require Digest::GOST };
eval { require Digest::GOST::CryptoPro };

my %digest = (
	'1' => ['Digest::SHA', 1],
	'2' => ['Digest::SHA', 256],
	'3' => ['Digest::GOST::CryptoPro'],
	'4' => ['Digest::SHA', 384],
	);

#
# source: http://www.iana.org/assignments/dns-sec-alg-numbers
#
{
	my @algbyname = (		## Reserved	=> 0,	# [RFC4034][RFC4398]
		'RSAMD5'	     => 1,			# [RFC3110][RFC4034]
		'DH'		     => 2,			# [RFC2539]
		'DSA'		     => 3,			# [RFC3755][RFC2536]
					## Reserved	=> 4,	# [RFC6725]
		'RSASHA1'	     => 5,			# [RFC3110][RFC4034]
		'DSA-NSEC3-SHA1'     => 6,			# [RFC5155]
		'RSASHA1-NSEC3-SHA1' => 7,			# [RFC5155]
		'RSASHA256'	     => 8,			# [RFC5702]
					## Reserved	=> 9,	# [RFC6725]
		'RSASHA512'	     => 10,			# [RFC5702]
					## Reserved	=> 11,	# [RFC6725]
		'ECC-GOST'	     => 12,			# [RFC5933]
		'ECDSAP256SHA256'    => 13,			# [RFC6605]
		'ECDSAP384SHA384'    => 14,			# [RFC6605]

		'INDIRECT'   => 252,				# [RFC4034]
		'PRIVATEDNS' => 253,				# [RFC4034]
		'PRIVATEOID' => 254,				# [RFC4034]
					## Reserved	=> 255,	# [RFC4034]
		);

	my %algbyval = reverse @algbyname;

	my @algbynum = map { ( $_, 0 + $_ ) } keys %algbyval;	# accept algorithm number

	my %algbyname = map { s /[^A-Za-z0-9]//g; $_ } @algbyname, @algbynum;


	sub algbyname {
		my $name = shift;
		my $key	 = uc $name;				# synthetic key
		$key =~ s /[^A-Z0-9]//g;			# strip non-alphanumerics
		return $algbyname{$key} || croak "unknown algorithm $name";
	}

	sub algbyval {
		my $value = shift;
		return $algbyval{$value} || $value;
	}
}

#
# source: http://www.iana.org/assignments/ds-rr-types
#
{
	my @digestbyname = (
		'SHA-1'	  => 1,					# RFC3658
		'SHA-256' => 2,					# RFC4509
		'GOST'	  => 3,					# RFC5933
		'SHA-384' => 4,					# RFC6605
		);

	my @digestbyalias = ( 'SHA' => 1 );

	my %digestbyval = reverse @digestbyname;

	my @digestbynum = map { ( $_, 0 + $_ ) } keys %digestbyval;    # accept algorithm number

	my %digestbyname = map { s /[^A-Za-z0-9]//g; $_ } @digestbyalias, @digestbyname, @digestbynum;


	sub digestbyname {
		my $name = shift;
		my $key	 = uc $name;				# synthetic key
		$key =~ s /[^A-Z0-9]//g;			# strip non-alphanumerics
		return $digestbyname{$key} || croak "unknown digest type $name";
	}

	sub digestbyval {
		my $value = shift;
		return $digestbyval{$value} || $value;
	}
}


sub decode_rdata {			## decode rdata from wire-format octet string
	my $self = shift;
	my ( $data, $offset ) = @_;

	my $length = $self->{rdlength} - 4;
	@{$self}{qw(keytag algorithm digtype digestbin)} = unpack "\@$offset n C2 a$length", $$data;
}


sub encode_rdata {			## encode rdata as wire-format octet string
	my $self = shift;

	return '' unless $self->{digtype};
	pack 'n C2 a*', @{$self}{qw(keytag algorithm digtype digestbin)};
}


sub format_rdata {			## format rdata portion of RR string.
	my $self = shift;

	return '' unless $self->{digtype};
	my @babble = BABBLE ? ( "\n;", $self->babble ) : ();
	my $digest = $self->digest;
	$digest = join( "\n", '(', split /(\S{64})/, $digest ) . ' )' if length $digest > 40;
	join ' ', @{$self}{qw(keytag algorithm digtype)}, $digest, @babble;
}


sub parse_rdata {			## populate RR from rdata in argument list
	my $self = shift;

	$self->$_(shift) for qw(keytag algorithm digtype);
	$self->digest(@_);
}


sub keytag {
	my $self = shift;

	$self->{keytag} = 0 + shift if scalar @_;
	return $self->{keytag} || 0;
}


sub algorithm {
	my ( $self, $arg ) = @_;

	unless ( ref($self) ) {		## class method or simple function
		my $argn = pop || croak 'undefined argument';
		return $argn =~ /[^0-9]/ ? algbyname($argn) : algbyval($argn);
	}

	return $self->{algorithm} unless defined $arg;
	return algbyval( $self->{algorithm} ) if $arg =~ /MNEMONIC/i;
	return $self->{algorithm} = algbyname($arg);
}


sub digtype {
	my ( $self, $arg ) = @_;

	unless ( ref($self) ) {		## class method or simple function
		my $argn = pop || croak 'undefined argument';
		return $argn =~ /[^0-9]/ ? digestbyname($argn) : digestbyval($argn);
	}

	return $self->{digtype} unless defined $arg;
	return digestbyval( $self->{digtype} ) if $arg =~ /MNEMONIC/i;
	return $self->{digtype} = digestbyname($arg);
}


sub digest {
	my $self = shift;

	$self->digestbin( pack "H*", map { die "!hex!" if m/[^0-9A-Fa-f]/; $_ } join "", @_ ) if scalar @_;
	unpack "H*", $self->digestbin() if defined wantarray;
}


sub digestbin {
	my $self = shift;

	$self->{digestbin} = shift if scalar @_;
	$self->{digestbin} || "";
}


sub babble {
	return Digest::BubbleBabble::bubblebabble( Digest => shift->digestbin ) if BABBLE;
	return '';
}


sub create {
	my $class = shift;
	my $keyrr = shift;
	my %args  = @_;

	my ($type) = reverse split '::', $class;

	my $kname = $keyrr->name;
	my $flags = $keyrr->flags;
	croak "Unable to create $kname $type record for non-DNSSEC key" unless $keyrr->protocol == 3;
	croak "Unable to create $kname $type record for NULL key" if ( $flags & 0xc000 ) == 0xc000;
	croak "Unable to create $kname $type record for key with flag bit7 clear" unless $flags & 0x0100;
	croak "Unable to create $kname $type record for key with flag bit6 set" if $flags & 0x0200;
	croak "Unable to create $kname $type record for key with flag bit0 set" if $flags & 0x8000;

	my $self = new Net::DNS::RR(
		name	  => $kname,				# per definition, same as keyrr
		type	  => $type,
		class	  => $keyrr->class,
		ttl	  => $keyrr->ttl,
		keytag	  => $keyrr->keytag,
		algorithm => $keyrr->algorithm,
		digtype	  => 1,					# SHA1 by default
		%args
		);

	my $owner = $self->{owner}->encode();
	my $data = pack 'a* a*', $owner, $keyrr->encode_rdata;

	my $arglist = $digest{$self->digtype} || croak 'unsupported digest type';
	my ( $object, @argument ) = @$arglist;
	my $hash = $object->new(@argument);
	$hash->add($data);
	$self->digestbin( $hash->digest );

	return $self;
}


sub verify {
	my ( $self, $key ) = @_;
	my $verify = create Net::DNS::RR::DS( $key, ( digtype => $self->digtype ) );
	return $verify->digestbin eq $self->digestbin;
}

1;
__END__


=head1 SYNOPSIS

    use Net::DNS;
    $rr = new Net::DNS::RR('name DS keytag algorithm digtype digest');

=head1 DESCRIPTION

Class for DNS Delegation Signer (DS) resource record.

=head1 METHODS

The available methods are those inherited from the base class augmented
by the type-specific methods defined in this package.

Use of undocumented package features or direct access to internal data
structures is discouraged and could result in program termination or
other unpredictable behaviour.


=head2 keytag

    $keytag = $rr->keytag;
    $rr->keytag( $keytag );

The 16-bit numerical key tag of the key. (RFC2535 4.1.6)

=head2 algorithm

    $algorithm = $rr->algorithm;
    $rr->algorithm( $algorithm );

Decimal representation of the 8-bit algorithm field.

algorithm() may also be invoked as a class method or simple function
to perform mnemonic and numeric code translation.

=head2 digtype

    $digtype = $rr->digtype;
    $rr->digtype( $digtype );

Decimal representation of the 8-bit digest type field.

digtype() may also be invoked as a class method or simple function
to perform mnemonic and numeric code translation.

=head2 digest

    $digest = $rr->digest;
    $rr->digest( $digest );

Hexadecimal representation of the digest over the label and key.

=head2 digestbin

    $digestbin = $rr->digestbin;
    $rr->digestbin( $digestbin );

Binary representation of the digest over the label and key.

=head2 babble

    print $rr->babble;

The babble() method returns the 'BubbleBabble' representation of the
The babble() method returns the 'BubbleBabble' representation of the
digest if the Digest::BubbleBabble package is available, otherwise
an empty string is returned.

BubbleBabble represents a message digest as a string of plausible
words, to make the digest easier to verify.  The "words" are not
necessarily real words, but they look more like words than a string
of hex characters.

The 'BubbleBabble' string is appended as a comment to the RDATA when
the string method is called.

=head2 create

    use Net::DNS::SEC;

    $dsrr = create Net::DNS::RR::DS($keyrr, digtype => 'SHA-256' );
    $keyrr->print;
    $dsrr->print;

This constructor takes a key object as argument and will return the
corresponding DS RR object.

The digest type defaults to SHA-1.

=head2 verify

    $verify = $dsrr->verify($keyrr);

The boolean verify method will return true if the hash over the key
RR provided as the argument conforms to the data in the DS itself
i.e. the DS points to the DNSKEY from the argument.


=head1 COPYRIGHT

Copyright (c)2001-2005 RIPE NCC.  Author Olaf M. Kolkman <olaf@net-dns.org>

Portions Copyright (c)2013 Dick Franks.

All Rights Reserved

Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
provided that the above copyright notice appear in all copies and that
both that copyright notice and this permission notice appear in
supporting documentation, and that the name of the author not be used
in advertising or publicity pertaining to distribution of the software
without specific prior written permission.

THE AUTHOR DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS; IN NO
EVENT SHALL AUTHOR BE LIABLE FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL
DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS
ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF
THIS SOFTWARE.

Package template (c)2009,2012 O.M.Kolkman and R.W.Franks.


=head1 SEE ALSO

L<perl>, L<Net::DNS>, L<Net::DNS::RR>, RFC4034, RFC3658

L<Algorithm Numbers|http://www.iana.org/assignments/dns-sec-alg-numbers>,
L<Digest Types|http://www.iana.org/assignments/ds-rr-types>

=cut
